### PR TITLE
Script to copy athena jdbc jar to our maven repo

### PR DESCRIPTION
**Pull request by snoe**
Since amazon is still not putting their athena jdbc on to maven, the easiest thing for us to do is setup a maven-y s3 repo and copy their jar into it.

This script allows you to run:
`./bin/athena.bb <version> [--deploy]`

For example:
`./bin/athena.bb 2.0.35 --deploy`

Will download the 2.0.35 jar from amazon's release bucket, create and update the necessary maven repo files, then upload all those files to our maven repo.

It uses the jar build WITHOUT the AWS SDK.

You must have `bb` (babashka), `aws` (awscli), and `shasum` commands available. You must have aws credentials available, e.g. through `aws configure`

### CODE REVIEW

File: bin/athena.bb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      z
+
+      :else
+      (recur (z/next z)))))
+
+(defn topmost [z]
+  (loop [z z]
+    (if-let [parent (z/up z)]
+      (recur parent)
+      z)))
```
Comment (metamben) [inline bin/athena.bb:43]: In this program, isn't this the same as
```suggestion
(def topmost (comp z/xml-zip z/root)
```

File: bin/athena.bb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+(defn jar-file-name [version]
+   (str "athena-jdbc-" version ".jar"))
+
+(defn create-new-metadata [version]
+  (let [existing (x/parse (io/reader "https://s3.amazonaws.com/maven-athena/com/metabase/athena-jdbc/maven-metadata.xml"))
```
Comment (metamben) [inline bin/athena.bb:49]: What if this file cannot be found?

File: bin/athena.bb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        z (z/xml-zip existing)]
+    (when (find-pred z (every-pred (comp #(= :version %) :tag)
+                                   (comp #(= version %) str/join :content)))
+      (throw (Exception. "Version already exists in maven-metadata.xml")))
+    (with-open [w (io/writer "maven-metadata.xml")]
+      (-> z
+          (find-pred (comp #(= :versions %) :tag))
```
Comment (metamben) [inline bin/athena.bb:56]: I wonder if `<versions>` can only ever occur in a single place, i.e., how well defined it is what node we edit.

File: bin/athena.bb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+          (z/insert-child (x/element :version {} version))
+          (topmost)
```
Comment (metamben) [inline bin/athena.bb:58]:
```suggestion
          topmost
```

**Review by metamben — APPROVED** [review]
My only real concern is why this script should fail if `https://s3.amazonaws.com/maven-athena/com/metabase/athena-jdbc/maven-metadata.xml` is missing.

The way the XML file is edited is not very robust, but it should be OK, for this script.

File: bin/athena.bb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+;; Will download the 2.0.35 jar from amazon's release bucket, create and update the necessary maven repo files, then upload all those files to our maven repo.
+;;
+;; It uses the jar build WITHOUT the AWS SDK.
+;;
+;; You must have bb (babashka), aws (awscli), and shasum commands available. You must have aws credentials available, e.g. through aws configure
+
+(defn find-pred [z p?]
+  (->> (iterate z/next z)
+       (take-while (every-pred some? (complement z/end?)))
+       (some #(when (-> % z/node p?) z))))
```
Comment (metamben) [inline bin/athena.bb:30]:
```suggestion
       (some #(when (-> % z/node p?) %))))
```

File: bin/athena.bb (hunk 2 above)
Comment (snoe) [reply bin/athena.bb:42]: You get an exception like this, which I figured is good enough for failing fast.

```
----- Error --------------------------------------------------------------------
Type:     java.io.FileNotFoundException
Message:  https://s3.amazonaws.com/maven-athena/com/metabase/athena-jdbc/maven-metadata.xml
```

Comment (metamben) [reply bin/athena.bb:42]: Wouldn't it be easy to bootstrap this? Does creating new metadata only make sense if this is not the first time?

**Pull request merged**